### PR TITLE
Use npm@3.5.2 to manage npm dependencies of Meteor packages.

### PR DESCRIPTION
**Pull request by benjamn**
Among many other improvements, the latest version of `npm` is especially good at flattening the structure of `node_modules` directories so that shared dependencies are (as much as possible) installed in only one location.

We already use `npm@3.1.2` to install the contents of `dev_bundle/{lib,server-lib}/node_modules`, with care to restore the original `npm@1.4.28` before bundling. This commit only uses `npm@1.4.28` to install the latest version of `npm` in place of the original version, and then uses `npm@3.3.9` for everything.

Getting `meteor test-packages` to run and pass required some very minor changes in `tools/isobuild/meteor-npm.js`, specifically not deleting the `package.json` and `node-shrinkwrap.json` files that we create in `.npm/package` directories.

It makes sense that `npm` would be so backwards compatible, since it is constrained to deliver the same files for a particular version of a package as it ever did. All the same, this is a potentially risky change, so I wanted to give it some time to bake in a pull request.

The next step in my testing plan is to install as many of the most popular Atmosphere packages as I can, to make sure this upgrade doesn't break their `npm install` process.

cc @martijnwalraven @stubailo @yyx990803 @glasser

### CODE REVIEW

File: tools/isobuild/meteor-npm.js [hunk 1]
```diff
@@ -258,9 +258,10 @@ var updateExistingNpmDirectory = function (packageName, newPackageNpmDir,
     // `npm install`
     installFromShrinkwrap(newPackageNpmDir);
 
-    // delete package.json and npm-shrinkwrap.json
-    files.unlink(files.pathJoin(newPackageNpmDir, 'package.json'));
-    files.unlink(files.pathJoin(newPackageNpmDir, 'npm-shrinkwrap.json'));
+    // Back when we used npm@1.4.28, we deleted the package.json and
+    // npm-shrinkwrap.json files created above, but npm@3.3.9 needs them
+    // to remain undeleted to avoid those pesky "couldn't read npm version
+    // lock information" errors.
```
Comment (glasser) [inline tools/isobuild/meteor-npm.js]: package.json still gets deleted (renamed) later in completeNpmDirectory, right?  And this is still OK?

(Not saying it should or shouldn't, but if you are leaving package.json around permanently (which it doesn't seem like you are) you should update the generated README to mention it)

**Comment by benjamn** [comment]
These 109 Atmosphere packages installed successfully using `meteor add`, and the app started up without any `npm`-related errors that I could see:

* `iron:router`
* `twbs:bootstrap`
* `aldeed:collection2`
* `momentjs:moment`
* `fourseven:scss`
* `meteorhacks:kadira`
* `tmeasday:publish-counts`
* `alanning:roles`
* `aldeed:autoform`
* `fortawesome:fontawesome`
* `mdg:geolocation`
* `meteorhacks:npm`
* `ian:accounts-ui-bootstrap-3`
* `dburles:collection-helpers`
* `kadira:flow-router`
* `meteorhacks:fast-render`
* `velocity:html-reporter`
* `aldeed:simple-schema`
* `matb33:collection-hooks`
* `sanjo:jasmine`
* `tap:i18n`
* `sacha:spin`
* `cfs:standard-packages`
* `percolate:synced-cron`
* `kadira:blaze-layout`
* `mdg:camera`
* `meteorhacks:aggregate`
* `raix:handlebar-helpers`
* `mike:mocha`
* `percolate:velocityjs`
* `nemo64:bootstrap`
* `materialize:materialize`
* `meteorhacks:subs-manager`
* `percolate:wireframing`
* `msavin:mongol`
* `flemay:less-autoprefixer`
* `meteorhacks:kadira-profiler`
* `mquandalle:jade`
* `aldeed:template-extension`
* `multiply:iron-router-progress`
* `cfs:gridfs`
* `dburles:google-maps`
* `aldeed:tabular`
* `meteoric:ionicons-sass`
* `xolvio:cucumber`
* `meteoric:ionic-sass`
* `u2622:persistent-session`
* `angularui:angular-ui-router`
* `meteoric:ionic`
* `meteorhacks:async`
* `semantic:ui`
* `rajit:bootstrap3-datepicker`
* `percolate:migrations`
* `natestrauser:font-awesome`
* `velocity:core`
* `urigo:angular-blaze-template`
* `manuelschoebel:ms-seo`
* `kadira:react-layout`
* `xolvio:template-isolator`
* `meteorhacks:unblock`
* `mquandalle:bower`
* `peerlibrary:aws-sdk`
* `cfs:filesystem`
* `djedi:sanitize-html`
* `mizzao:bootboxjs`
* `reywood:publish-composite`
* `awatson1978:browser-detection`
* `tsega:bootstrap3-datetimepicker`
* `meteorhacks:meteor-down`
* `froatsnook:sleep`
* `doctorpangloss:filter-collections`
* `aldeed:http`
* `kadira:blaze-plus`
* `abhiaiyer:meteor-twilio`
* `doctorpangloss:method-hooks`
* `sergeyt:typeahead`
* `velocity:meteor-stubs`
* `copleykj:livestamp`
* `jparker:crypto-md5`
* `zimme:active-route`
* `jparker:crypto-base64`
* `zimme:iron-router-active`
* `useraccounts:bootstrap`
* `meteorhacks:ssr`
* `cosmos:browserify`
* `jparker:crypto-core`
* `yogiben:autoform-file`
* `rubaxa:sortable`
* `cfs:graphicsmagick`
* `monbro:mongodb-mapreduce-aggregation`
* `perak:user-roles`
* `stevezhu:lodash`
* `juliancwirko:s-alert`
* `driftyco:ionic`
* `harrison:papa-parse`
* `natestrauser:animate-css`
* `meteortoys:allthings`
* `percolate:momentum`
* `matteodem:easy-search`
* `angular:angular-material`
* `mrgalaxy:stripe`
* `aslagle:reactive-table`
* `yogiben:admin`
* `meteorhacks:cluster`
* `jparker:gravatar`
* `edgee:slingshot`
* `angularutils:pagination`
* `okgrow:analytics`
* `mizzao:user-status`

File: tools/isobuild/meteor-npm.js (hunk 1 above)
Comment (benjamn) [inline tools/isobuild/meteor-npm.js]: After looking at this code some more, I think we can keep deleting those files as long as we do it after the call to `getInstalledDependencies` (which runs the failing `npm ls --json` command). I'll push another commit momentarily.

Comment (benjamn) [inline tools/isobuild/meteor-npm.js]: Pushed: https://github.com/meteor/meteor/commit/c4f774f452d435ba0830cea814f3e4eccb9d3366

**Comment by stubailo** [comment]
> These 109 Atmosphere packages installed successfully using meteor add, and the app started up without any npm-related errors that I could see:

It seems that the real test would be to see if these packages can be built, right? Because the NPM dependencies are baked in to the published bundle already.

**Comment by glasser** [comment]
The real test (for me :) ) is less about "can these be built" and more about "does our custom shrinkwrap-surgery code still work" — what happens if you start with some package (old or new), and edit versions of dependencies, add and remove dependencies, edit versions only mentioned in shrinkwrap, etc.  (eg: if you have Npm.depends(bar) and bar -> foo, and you edit the now-top-level foo version in npm-shrinkwrap.json, does meteor properly detect that and rebuild with the right version?)

If the answer ends up being "no, it doesn't, and we have to do some work to fix it"... I would highly suggest we try to figure out if we can come up with a way to make the code less clever (if perhaps less powerful) by not doing shrinkwrap surgery at all...  for example, I suspect that if we just made meteor-npm run `npm install --save --save-exact` instead of doing shrinkwrap manipulation ourselves, it would Just Work.

**Comment by benjamn** [comment]
@glasser I totally agree with that decision tree, though I'm already pretty sure rewriting `meteor-npm.js` falls within the scope of this pull request. For example, you can delete the entire directory for an indirect dependency, and nothing about the old logic or the new logic will trigger an update.

Looking ahead a bit, my hope is to install package dependencies in a common `node_modules` directory, so that `npm@3` can manage dependencies shared between different Meteor packages. In that bright future, I'm optimistic that the burden of fixing missing or corrupted dependencies can be shouldered entirely by `npm`.

**Comment by glasser** [comment]
@benjamn That sounds great! If you need any help understanding the code in the tool that manages lists of node modules (it's annoying pervasive and kinda confusing) let me know.

I suspect that, just like Meteor 1.2 moved most plugin-level processing from "compiler" to "bundler", you could move most npm processing from "compiler" to "bundler" too, and I think that would simplify a lot.  ie, only run `npm` in the context of an app (as opposed to running it in the contexts of individual packages).

**Comment by benjamn** [comment]
After thinking a lot about merging package dependencies, and discussing it with @glasser yesterday, I think that's too ambitious for this pull request, and potentially not very worthwhile, since Meteor packages depend on exact versions of `npm` packages, so there probably wouldn't be very many opportunities to share direct dependencies, unless the versions were exactly the same.

I have also now taken some time to audit `tools/isobuild/meteor-npm.js`, and I believe that code works as well as it ever did using `npm@3.3.9`, as long as a single package's `npm` dependencies all end up in `.npm/package/node_modules/`, which they do.

In other words, I believe this pull request is potentially ready to merge, with the additional caveat that I will be working with this code extensively in the context of implementing modules, and will keep an eye out for anything broken as a result of this upgrade.

**Comment by glasser** [comment]
Have you done the specific test I suggested where you edit the shrinkwrap file's version of an indirect dependency?

**Comment by benjamn** [comment]
@glasser yes, that test works. For example, the `http` package has a bunch of indirect `npm` dependencies via the `request` package, and whenever I change its `npm-shrinkwrap.json` file, the server restarts with `http: updating npm dependencies -- request...`, and the original contents of the shrinkwrap file are restored.

**Comment by glasser** [comment]
Er, when you say "the original contents of the shrinkwrap file are restored", do you mean your manual changes are reverted?  Because that's what I'd consider a test failure.  ie, if you `git pull` and the only change is to change an indirect dependency, then `meteor-npm` should update your local install to match that change, not revert the change that you pulled.

**Comment by benjamn** [comment]
Got it, I'll keep looking into that.

Another problem I'm concerned about is what happens when you rebuild an app that was previously built without these changes:
```sh
cd path/to/meteor/checkout
git checkout devel
find packages | grep '\.npm$' | xargs rm -rf
cd /tmp
meteor create test
cd test
meteor # This works as usual.
```
Now, if you switch to the `npm3` branch and run the app again:
```sh
cd path/to/meteor/checkout
git checkout -t origin/npm3
cd /tmp/test
meteor
```
Then you get errors like
```sh
While building package npm-mongo:
error: couldn't read npm version lock information: Command failed: npm ERR! invalid: readable-stream@2.0.2 /Users/ben/meteor/packages/npm-mongo/.npm/package/node_modules/mongodb/node_modules/readable-stream
npm ERR! invalid: readable-stream@2.0.2 /Users/ben/meteor/packages/npm-mongo/.npm/package/node_modules/mongodb/node_modules/readable-stream
```

**Comment by benjamn** [comment]
Meteor 1.3 upgraded `npm` to version 2.14.22, which is a big improvement. Upgrading to `npm@3` will have to wait until I find more time.
